test/ResourceSourceICUUnexplainedParams.test.js

```diff
@@ -0,0 +1,205 @@
+        const resource = new ResourceString({
+            key: "icu.test",
+            sourceLocale: "en-US",
+            source: "{count, plural, one {{featuresLink} about our new feature.} other {{featuresLink} about all our new features.}}",
+            pathName: "a/b/c.xliff",
+            comment: "Prompt about one or more new features."
+        });
+
+        const result = rule.matchString({
+            source: /** @type {string} */ (resource.getSource()),
+            resource,
+            file: "a/b/c.xliff"
+        });
+
+        const expected = [
+            new Result({
+                severity: "warning",
+                description: `Replacement parameter "featuresLink" is not mentioned in the string's comment for translators.`,
+                id: "icu.test",
+                source: "{count, plural, one {{featuresLink} about our new feature.} other {{featuresLink} about all our new features.}}",
+                highlight:
+                    "{count, plural, one {<e0>{featuresLink}</e0> about our new feature.} other {{featuresLink} about all our new features.}}",
+                rule,
+                pathName: "a/b/c.xliff"
+            }),
+            new Result({
+                severity: "warning",
+                description: `Replacement parameter "featuresLink" is not mentioned in the string's comment for translators.`,
+                id: "icu.test",
+                source: "{count, plural, one {{featuresLink} about our new feature.} other {{featuresLink} about all our new features.}}",
+                highlight:
+                    "{count, plural, one {{featuresLink} about our new feature.} other {<e0>{featuresLink}</e0> about all our new features.}}",
+                rule,
+                pathName: "a/b/c.xliff"
+            })
+        ];
+
+        expect(result).toStrictEqual(expected);
+    });
+
+    test("no comment for translators provided", () => {
+        const resource = new ResourceString({
+            key: "icu.test",
+            sourceLocale: "en-US",
+            source: "{featuresLink} about the new features.",
+            pathName: "a/b/c.xliff"
+        });
+
+        const result = rule.matchString({
+            source: /** @type {string} */ (resource.getSource()),
+            resource,
+            file: "a/b/c.xliff"
+        });
+
+        expect(result).toEqual(undefined);
+    });
+
+    test("no source string available in the resource", () => {
+        const resource = new ResourceString({
+            key: "icu.test",
+            sourceLocale: "en-US",
+            source: undefined,
+            pathName: "a/b/c.xliff"
+        });
+
+        const result = rule.matchString({
+            source: /** @type {string} */ (resource.getSource()),
+            resource,
+            file: "a/b/c.xliff"
+        });
+
+        expect(result).toEqual(undefined);
+    });
+});
```

Comment:
Add tests:

1. Where the comment explains the param, but uses a different case
2. Where the comment explains the param but uses a colon, quotes, or other punctuation around it. Example: "count: the number of rules", "'count' refers to the number of rules", "{count} is the number of rules", and "\<count> is the number of rules" should all be fine and not produce a result.
4. Where the comment does not explain the param, but the param name is part of a random other word in the comment.
5. Where the comment explains a param that doesn't exist in the source
6. Where the comment does not explain the param used for the pivot of a plural or select. (ie. where source is {count, plural, ...} and the comment does not explain "count".)
7. Where the resource type is array
8. Where the resource type is plural

Reply:
1, 2, 3 in cbbde77

Reply:
6, 7 in 2c406c2

Reply:
4: what's the point? Unless we wish to introduce some set structure for the comment or a requirement to specifically mark a replacement parameter somehow (brackets, backticks), there is hardly a way to tell that a description for a non-existent replacement parameter was provided. Effectively, such test is the same as the base `replacement parameter not mentioned in comment for translators`.

Reply:
5: I thought we only want to flag `ArgumentElement`s. Should I also add support for tag element then?

Reply:
5 in 7fecf38 (added support for selects, plurals and tag elements)

Reply:
plural, select, and selectordinal are just special cases of the general replacement parameter syntax:

```
{variable, how-to-format-the-variable, options-for-the-formatting}
```

where the "how to format" and the "formatting options" can be left out if there are none.

for example:

```
Price for {purchasedItem} is {price, number, ::currency/USD} with {discountPct, number, percent} discount.
```

For plurals, we need to check the special case of the # replacement param:

```
{count, plural, one {There is # item.} other {There are # items.}}
```

In this case, we recursively find replacement params, so when we get to the choices, we find the # character. We need to remember that the pivot variable for the whole select/plural is "count", so we need to check if "count" is mentioned in the comments. If the plural category string uses "{count}" directly, then it is easy -- we just check as normal. It is the "#" that is a special case.

Tags, however, refer to xml tags in the string and do not need the same explanation as replacement parameters.

Reply:
So what do you think should be different here?

To avoid confusion, I'm going to use naming from here https://formatjs.io/docs/core-concepts/icu-syntax/#formatted-argument i.e.:
- `{key}` is simple argument
- `{key, type, format}` is formatted argument
- for type `plural | select | selectOrdinal`, format is `matches` and each of them is `match {output}`
- `text<tag>text</tag>text` is rich text formatting
Additionally, I'm going to call `#` within the plural or selectOrdinal outputs a `pound` per [parser library typedef](https://github.com/formatjs/formatjs/blob/58d6a7b398d776ca3d2726d72ae1573b65cc3bef/packages/icu-messageformat-parser/types.ts#L41) (even though I always thought this is a hash and pound is £).

1. select & plural key warning: initially I figured that we are not going to search for them, because the match output usually provides enough context on its own. It's implemented now in 7fecf38
2. selectOrdinal: from [parser typedef](https://github.com/formatjs/formatjs/blob/58d6a7b398d776ca3d2726d72ae1573b65cc3bef/packages/icu-messageformat-parser/types.ts#L8) it looks like selectOrdinal is conflated either with select or plural (not sure which) - either way, the rule will also search for its key since 7fecf38
3. _For plurals, we need to check the special case of the # replacement param_: I don't understand what you mean - do you think we should search whether the pound symbol explicitly appears in the translation's comment and warn if it's absent?
4. _If the plural category string uses "{count}" directly, then it is easy -- we just check as normal._: that's already implemented. The whole message is being traversed recursively, and within the match output `{count}` messge parser treats it as a simple argument (as opposed to the pound, which according to the docs is equivalent to `{count, number}`)
5. _Tags, however, refer to xml tags in the string and do not need the same explanation_: [per FormatJS docs](https://formatjs.io/docs/core-concepts/icu-syntax/#rich-text-formatting) they don't have to be XML tags exactly, so after it turned out we would also like to search for formatted argument keys, I figured we might also want to search for rich formatting tags as well (like `<cta></cta>` in [this example](https://formatjs.io/docs/react-intl/components/#rich-text-formatting)). It's easy to disable though, let me know if I should do that